### PR TITLE
logrotate: general improvements

### PR DESCRIPTION
**Pull request by dougsland**

### CODE REVIEW

**Comment by dougsland** [comment]
Initial example:

$ cat /etc/redhat-release 
Fedora release 24 (Twenty Four)

$ ./autogen.sh && ./configure 
$  rpmbuild -ba logrotate.spec 
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.P7DzOV
+ umask 022
+ cd /home/github/rpmbuild/BUILD
+ cd /home/github/rpmbuild/BUILD
+ rm -rf logrotate-3.10.0
+ /usr/bin/gzip -dc /home/github/rpmbuild/SOURCES/logrotate-3.10.0.tar.gz
+ /usr/bin/tar -xof -
+ STATUS=0
+ '[' 0 -ne 0 ']'
+ cd logrotate-3.10.0
+ /usr/bin/chmod -Rf a+rX,u+w,g-w,o-w .
+ exit 0
Executing(%build): /bin/sh -e /var/tmp/rpm-tmp.1A1WP4
+ umask 022
+ cd /home/github/rpmbuild/BUILD
+ cd logrotate-3.10.0
+ make -j4 'RPM_OPT_FLAGS=-O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -m64 -mtune=generic' WITH_SELINUX=yes
make: *** No targets specified and no makefile found.  Stop.
error: Bad exit status from /var/tmp/rpm-tmp.1A1WP4 (%build)

**Comment by dougsland** [comment]
Additionally, just for a note to %make_install patch, the correct prefix for make install is:
make prefix=/mydir and not make PREFIX=/mydir

Example:
$ make prefix=~/tmp2 install
Making install in .
make[1]: Entering directory '/home/github/logr/logrotate'
make[2]: Entering directory '/home/github/logr/logrotate'
 /usr/bin/mkdir -p '/home/github/tmp2/sbin'
  /usr/bin/install -c logrotate '/home/github/tmp2/sbin'
<snip>

$ make PREFIX=~/tmp2 install
Making install in .
make[1]: Entering directory '/home/github/logr/logrotate'
make[2]: Entering directory '/home/github/logr/logrotate'
 /usr/bin/mkdir -p '/usr/local/sbin'
  /usr/bin/install -c logrotate '/usr/local/sbin'
/usr/bin/install: cannot create regular file '/usr/local/sbin/logrotate': Permission denied
Makefile:411: recipe for target 'install-sbinPROGRAMS' failed
<snip>

More info:
https://www.gnu.org/software/make/manual/html_node/Directory-Variables.html

**Comment by glensc** [comment]
@dougsland you're doing make install wrong.

you should be doing `make install DESTDIR=$RPM_BUILD_ROOT`

example spec: https://github.com/pld-linux/logrotate/blob/auto/th/logrotate-3.10.0-1/logrotate.spec

**Comment by glensc** [comment]
@dougsland seems you missed [CONTRIBUTING.md](https://github.com/logrotate/logrotate/blob/master/CONTRIBUTING.md) point telling not to use `master` branch and pushed last two commits accidentally to this PR?

to fix this up (and keep your own developments on other branch than `master` until this PR gets merged)
- git reset --hard <last_good_commit>
- git push -f

**Comment by glensc** [comment]
err, the downvote was for c1f43b5a48dfff49f3a0a3efb4481fdc5f9f9fba commit only. don't know how to edit "github review comments"

**Comment by dougsland** [comment]
Hi Elan,

Thanks for your input, additional data below.

>  @dougsland you're doing make install wrong.
>  you should be doing make install DESTDIR=$RPM_BUILD_ROOT
> example spec: https://github.com/pld-linux/logrotate/blob/auto/th/logrotate-3.10.0-1/logrotate.spec

I agree that we should, at least, replace with DESTDIR, my example was just to show it's broken in master. However, I have used the macro instead of the full syntax. I am not sure which distro versions logrotate in *master* branch should support but for old distros (Centos 4 or 5 and others variants) we can always if required backport it correctly for an specific branch. If you guys prefer the explicit DESTDIR in master, I can change it, just let me know the consensus.

# cat /etc/os-release
<snip>
ID=opensuse
ID_LIKE="suse"
VERSION_ID="20161006"
</snip>
# rpm --eval '%{make_install}'
make install DESTDIR=/home/linux/rpmbuild/BUILDROOT/%{name}-%{version}-%{release}.x86_64

# cat /etc/redhat-release 
CentOS release 6.8 (Final)
# rpm --eval %{make_install}
make install DESTDIR=/root/rpmbuild/BUILDROOT/%{name}-%{version}-%{release}.x86_64

# cat /etc/redhat-release 
CentOS Linux release 7.1.1503 (Core) 
[root@localhost ~]# rpm --eval %{make_install}
/usr/bin/make install DESTDIR=/root/rpmbuild/BUILDROOT/%{name}-%{version}-%{release}.x86_64

# cat /etc/redhat-release 
Fedora release 24 (Twenty Four)
#  rpm --eval %{make_install}
/usr/bin/make install DESTDIR=/root/rpmbuild/BUILDROOT/%{name}-%{version}-%{release}.x86_64

From: https://fedoraproject.org/wiki/Packaging:Guidelines#Macros
"""
Fedora packages should use: %make_install (Note the "_" !), make DESTDIR=%{buildroot} install or make DESTDIR=$RPM_BUILD_ROOT install. Those all do the same thing. 
"""

>
> example spec: https://github.com/pld-linux/logrotate/blob/auto/th/logrotate-3.10.0-1/logrotate.spec
>
humm, why your change isn't in master or in pull request (At least, I can't see it)


> @dougsland seems you missed CONTRIBUTING.md point telling not to use master branch and 
> pushed last two commits accidentally to this PR?
> 
> to fix this up (and keep your own developments on other branch than master until this PR gets 
> merged)
>
>    git reset --hard
>    git push -f

Thanks, I have missed the CONTRIBUTING.md, I will fix the mess.

**Comment by glensc** [comment]
> Q: example spec: https://github.com/pld-linux/logrotate/blob/auto/th/logrotate-3.10.0-1/logrotate.spec
> humm, why your change isn't in master or in pull request (At least, I can't see it)

Several reasons actually:
- logrotate has not always been in github, and reception of bug reports was really poor before that
- it is impossible to write generic .spec that fits for all distributions. history has shown to me that it's not worth to send .spec changes to upstream, upstream has different priorities and standards
- my distro uses only the spec i linked, so i was not even aware that upstream has .spec and it is broken.

**Comment by glensc** [comment]
regarding `%make_install`: i do not know how old the macro is, however i commented that you used `prefix=` and suggested to use `destdir=`, i did not pay close attention.

anyway, here's `%make_install` state from my distro(s):
```
$ lsb_release -irc
Distributor ID: PLD
Release:        3.0
Codename:       Th
$ rpm --eval %{make_install}
make install DESTDIR=/home/users/glen/tmp/%{name}-%{version}-root-glen
```

```
$ lsb_release -irc
Distributor ID: PLD
Release:        2.0
Codename:       Ac

$ rpm --eval %{make_install}
%{make_install}
```

so pld ac (rpm -4.5) does not have the macro. but it's irrelevant because:
- logrotate is not going to be updated there
- distro has own .spec to build logrotate from

File: logrotate.c [hunk 1]
```diff
@@ -607,16 +607,16 @@ static int compressLogFile(char *name, struct logInfo *log, struct stat *sb)
...
+    while ((i = read(compressPipe[0], buff, sizeof(buff) - 1)) > 0) {
+            if (!error_printed) {
+                error_printed = 1;
+		message(MESS_ERROR, "Compressing program wrote following message "
+                               "to stderr when compressing log %s:\n", name);
+            }
+            buff[i] = '\0';
+            fprintf(stderr, "%s", buff);
+    }
+    close(compressPipe[0]);
```
Comment (kdudka) [inline logrotate.c]: This hunk actually breaks the indentation (random tabs replaced by spaces).  I will tweak the commit not to do so.

File: logrotate.c [hunk 2]
```diff
@@ -2523,6 +2524,14 @@ int main(int argc, const char **argv)
     logSetLevel(MESS_NORMAL);
     setlocale (LC_ALL, "");
 
+    if (stat(STATEDIR, &st) == -1) {
+        mkdir(STATEDIR, 0755);
+    }
+
+    if (stat(OLDSTATEFILE, &st) == 0) {
+        unlink(OLDSTATEFILE);
```
Comment (kdudka) [inline logrotate.c]: This introduces a TOCTOU race.  Is not it better to just call ```mkdir()```?  It fails with ```EEXISTS``` if the target already exists.

Comment (kdudka) [inline logrotate.c]: I do not think that existing state file should be silently removed.  We should move it to the new location instead.

**Pull request closed without merging**